bibliopixel/animation/base.py

```diff
@@ -133,55 +117,60 @@ def _run(self, amt, fps, sleep, max_steps, untilComplete, max_cycles, seconds):
                     "Frame: %sms / Update Max: %sms", stepTime, updateTime)
             else:
                 log.debug("%sms/%sfps / Frame: %sms / Update: %sms",
-                          totalTime, int(1000 / max(totalTime, 1)), stepTime, updateTime)
+                          totalTime, int(1000 / max(totalTime, 1)), stepTime,
+                          updateTime)
 
             if sleep:
                 diff = (self._msTime() - self._timeRef)
                 t = max(0, (sleep - diff) / 1000.0)
                 if t == 0:
                     log.warning(
                         "Frame-time of %dms set, but took %dms!", sleep, diff)
-                if self._threaded:
+                if self._led._threadedAnim:
                     self._stopEvent.wait(t)
                 else:
                     time.sleep(t)
             cur_step += 1
 
-        self.__exit__(None, None, None)
+        self.cleanup()
+        self._callback and self._callback(self)
 
-        if self._callback:
-            self._callback(self)
+    def run(self, amt=1, fps=None, sleep=None, max_steps=0, until_complete=False,
+            max_cycles=0, threaded=False, joinThread=False, callback=None,
+            seconds=None):
 
-    def run(self, amt=1, fps=None, sleep=None, max_steps=0, untilComplete=False, max_cycles=0, threaded=False, joinThread=False, callback=None, seconds=None):
-
-        self._led._threadedAnim = self._threaded = threaded
-        if self._threaded:
+        self._callback = callback
+        self._led._threadedAnim = threaded
+        if self._led._threadedAnim:
             self._stopEvent.clear()
         self._callback = callback
 
-        if self._threaded:
-            args = {}
-            l = locals()
-            run_params = ["amt", "fps", "sleep",
-                          "max_steps", "untilComplete", "max_cycles", "seconds"]
-            for p in run_params:
-                if p in l:
-                    args[p] = l[p]
+        def do_run():
+            self._run(
+                amt, fps, sleep, max_steps, until_complete, max_cycles, seconds)
```

Comment:
yes, looks bad... let me do something... Hah, I must have not liked it either, because it's cleaned up in a later commit!